test/functional/multifactor_auths_controller_test.rb

```diff
@@ -39,122 +105,447 @@ class MultifactorAuthsControllerTest < ActionController::TestCase
+            assert_nil @controller.session[:mfa_redirect_uri]
+          end
+        end
+
+        context "when session is expired" do
+          setup do
+            get :update, params: { level: "ui_and_api" }
+
+            travel 16.minutes do
+              put :mfa_update, params: { otp: ROTP::TOTP.new(@user.mfa_seed).now }
             end
           end
 
-          context "when otp code is incorrect" do
+          should redirect_to("the settings page") { edit_settings_path }
+
+          should "not update mfa level" do
+            assert_predicate @user.reload, :mfa_ui_only?
+          end
+
+          should "set flash error" do
+            assert_equal "Your login page session has expired.", flash[:error]
+          end
+
+          should "clear session variables" do
+            assert_nil @controller.session[:mfa_expires_at]
+            assert_nil @controller.session[:level]
+            assert_nil @controller.session[:mfa_redirect_uri]
+          end
+        end
+      end
+
+      context "on PUT to webauthn_update" do
+        setup do
+          put :update, params: { level: "ui_and_api" }
+          post :webauthn_update
+        end
+
+        should redirect_to("the settings page") { edit_settings_path }
+
+        should "set flash error" do
+          assert_equal "You don't have any security devices enabled. " \
+                       "You have to associate a device to your account first.", flash[:error]
+        end
+
+        should "not update mfa level" do
+          assert_predicate @user.reload, :mfa_ui_only?
+        end
+
+        should "clear session variables" do
+          assert_nil @controller.session[:mfa_expires_at]
+          assert_nil @controller.session[:level]
+          assert_nil @controller.session[:mfa_redirect_uri]
+        end
+      end
+    end
+
+    context "when a webauthn device is enabled" do
+      setup do
+        @webauthn_credential = create(:webauthn_credential, user: @user)
+      end
+
+      context "on POST to create totp mfa" do
+        setup do
+          @seed = ROTP::Base32.random_base32
+          @controller.session[:mfa_seed] = @seed
+          @controller.session[:mfa_seed_expire] = Gemcutter::MFA_KEY_EXPIRY.from_now.utc.to_i
+
+          perform_enqueued_jobs only: ActionMailer::MailDeliveryJob do
+            post :create, params: { otp: ROTP::TOTP.new(@seed).now }
+          end
+        end
+
+        should respond_with :success
+
+        should "keep mfa enabled" do
+          assert_predicate @user.reload, :mfa_enabled?
+        end
+
+        should "send totp enabled email" do
+          assert_emails 1
+          assert_equal "Multi-factor authentication enabled on RubyGems.org", last_email.subject
+          assert_equal [@user.email], last_email.to
+        end
+      end
+
+      context "on PUT to update mfa level" do
+        context "on updating to ui_and_api" do
+          setup do
+            freeze_time
```

Comment:
With the current setup, mfa is already enabled at `ui_and_api`. Should change the mfa level to test this case.